url3010dbde70_nbartlett_Sequence_Memoizer_tgz-pJ8-edu_columbia_stat_wood_pub_sequencememoizer_IntSequenceMemoizerJ8/src/edu/columbia/stat/wood/pub/sequencememoizer/util/IntRestaurant.java

```diff
@@ -0,0 +1,478 @@
+            customers++;
+            tables++;
+        } else if (type > types[types.length - 1]) {
+            p *= (double) customers / ((double) tables * discount);
+            sr.set(true, 0, customers, tables);
+
+            insertNewType(type, types.length);
+            customers++;
+            tables++;
+        } else {
+            int index;
+
+            index = getIndex(type);
+            if (types[index] != type) {
+                p *= (double) customers / ((double) tables * discount);
+                sr.set(true, 0, customers, tables);
+
+                insertNewType(type, index);
+                customers++;
+                tables++;
+            } else {
+                double numerator, denominator;
+                int tci, tti, tc, tt;
+
+                tci = index * 2;
+                tti = tci + 1;
+
+                tc = customersAndTables[tci];
+                tt = customersAndTables[tti];
+
+                numerator = (double) tc - (double) tt * discount;
+
+                p -= numerator / (double) customers;
+                p *= (double) customers / ((double) tables * discount);
+
+                denominator = numerator + (double) tables * discount * p;
+
+                if (numerator / denominator > IntSequenceMemoizer.RNG.nextDouble()) {
+                    sr.set(false, customersAndTables[tti], customers, tables);
+
+                    customersAndTables[tci]++;
+                    customers++;
+                } else {
+                    sr.set(true, customersAndTables[tti], customers, tables);
+
+                    customersAndTables[tci]++;
+                    customersAndTables[tti]++;
+                    customers++;
+                    tables++;
+                }
+            }
+        }
+
+        if (customers >= sm.maxCustomersInRestaurant) {
+            deleteCustomers((int) (sm.maxCustomersInRestaurant * .1),discount);
+        }
+
+        return p;
+    }
+
+    private void insertNewType(int type, int index) {
+        int[] newTypes;
+        int[] newCustomersAndTables;
+        int l;
+
+        l = types.length;
+        newTypes = new int[l + 1];
+        newCustomersAndTables = new int[2 * l + 2];
+
+        System.arraycopy(types, 0, newTypes, 0, index);
+        System.arraycopy(customersAndTables, 0, newCustomersAndTables, 0, index * 2);
+
+        newTypes[index] = type;
+        newCustomersAndTables[2 * index] = 1;
+        newCustomersAndTables[2 * index + 1] = 1;
+
+        System.arraycopy(types, index, newTypes, index + 1, l - index);
+        System.arraycopy(customersAndTables, index * 2, newCustomersAndTables, index * 2 + 2, 2 * l - 2 * index);
+
+        types = newTypes;
+        customersAndTables = newCustomersAndTables;
+    }
+
+    public IntRestaurant fragmentForInsertion(IntRestaurant irParent, int irEdgeStart, int irEdgeLength, IntSeqNode irEdgeNode, double discount, double irDiscount, MutableLong restCount) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

This method contains 44 lines of code, not including blank lines or lines with only comments, Java punctuation characters, identifiers, or literals. By comparison, 98% of the methods in the CodeGuru reference dataset contain fewer lines of code. Large methods might be difficult to read and have logic that is hard to understand and test. We recommend that you simplify this method or break it into multiple methods.